tests/test_salesforce_api_per_run_quota.py

```diff
@@ -0,0 +1,75 @@
+from sfbase import SFBaseTest
+from tap_tester import connections, runner, menagerie
+from tap_tester.logger import LOGGER
+
+
+class SFAPIQuota(SFBaseTest):
+    """
+    https://jira.talendforge.org/browse/TDL-23431
+    This testcase makes sure we are able to configure the tap with specific 
+    per_run quota for apis and verifies if tap is working as per the 
+    api quota limit set.
+    """
+
+    """
+    Set a start date well past so that we have enough data to sync to hit the limit
+    set a low quota to make the sync fail
+    some streams are excluded as they don't have any data 
+    """
+
+    start_date = '2000-11-23T00:00:00Z'
+    per_run_quota = '1'
+    streams_to_exclude = {
+             'DatacloudAddress',
+             'DatacloudCompany',
+             'DatacloudContact',
+             'DatacloudDandBCompany',
+             'DatacloudOwnedEntity',
+             'DatacloudPurchaseUsage',
+             'FieldSecurityClassification',
+             'ServiceAppointmentStatus',
+             'WorkOrderLineItemStatus',
+             'WorkOrderStatus',
+             'ShiftStatus',
+             'WorkStepStatus',
+        }
+
+    @staticmethod
+    def name():
+        return "tt_sf_api_quota"
+
+    def streams_to_test(self):
+        return self.expected_streams().difference( self.streams_to_exclude )
+
+    @staticmethod
+    def streams_to_selected_fields():
+        # Select all fields in the selected catalogs (streams)
+        return {}
+
+    def test_api_per_run_quota(self):
+        """
+        Run the tap in check mode and verify it returns the error for quota limit reached. If not, proceed to sync mode.
+        """
+        expected_per_run_error = "Terminating replication due to allotted quota"
+        expected_total_quota_error = "Terminating replication to not continue past configured percentage "
+        conn_id = connections.ensure_connection(self)
+
+        # run check mode
+        # Run a check job using orchestrator (discovery)
+        with self.assertRaises(Exception) as ex:
```

Comment:
per run did not raise exception on last commit.  could the check job use less than 1 percent quota?  

```
FAIL: test_api_per_run_quota (test_salesforce_api_per_run_quota.SFAPIQuota)
Run the tap in check mode and verify it returns the error for quota limit reached. If not, proceed to sync mode.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/project/tests/test_salesforce_api_per_run_quota.py", line 60, in test_api_per_run_quota
    found_catalogs = self.run_and_verify_check_mode( conn_id)
AssertionError: Exception not raised

2023-08-30 14:32:00,401 TapTester - INFO     ***** Test Results *****
----------------------------------------------------------------------
Ran 1 tests in 107s
```

Reply:
We can't have less than 1%, couldn't send fraction amount. This is very tricky, let's discuss.

Reply:
If it is not guaranteed to hit the per run quota in check mode can we do a "try / catch" for check mode and then use the normal "assertRaises" for sync mode?  That way if the quota is hit during check mode we can catch the error and verify it and if it is not hit until sync mode we can also catch it there.  If that works I'd use the same pattern for both tests.